libraries/llvm/io.c

```diff
@@ -277,289 +246,155 @@ int uv_error_to_errno(int uv_err) {
+typedef struct Listeners {
+    Stack head;
+    struct Listeners* tail;
+} Listeners;
 
-    // Load the callbacks
-    Callbacks* callbacks = (Callbacks*)req->data;
-    struct Neg success = callbacks->on_success;
-    struct Neg failure = callbacks->on_failure;
+typedef struct {
+    uint64_t rc;
+    void* eraser;
+    promise_state_t state;
+    // state of {
+    //   case UNRESOLVED => Possibly empty (head is NULL) list of listeners
+    //   case RESOLVED   => Pos (the result)
+    // }
+    union {
+        struct Pos value;
+        Listeners listeners;
+    } payload;
+} Promise;
 
-    // Free request structure
-    uv_fs_req_cleanup(req);
-    free(req);
-    free(callbacks);
+void c_promise_erase_listeners(void *envPtr) {
+	struct {promise_state_t state; union {struct Pos value; Listeners listeners;} payload;} *valPtr = envPtr;
+	promise_state_t state = valPtr->state;
 
-    if (result >= 0) {
-        eraseNegative(failure);
-        run_i64(success, result);
-    } else {
-        eraseNegative(success);
-        run_i64(failure, uv_error_to_errno(result));
+    switch (state) {
+        case UNRESOLVED:
+			Stack head = valPtr->payload.listeners.head;
+			Listeners* tail = valPtr->payload.listeners.tail;
+			if (head != NULL) {
+			    // Erase head
+				eraseStack(head);
+				// Erase tail
+				Listeners* current = tail;
+				while (current != NULL) {
+					Stack head = current->head;
+					Listeners* tail = current->tail;
+					free(current);
+					eraseStack(head);
+					current = tail;
+				};
+			};
+            break;
+        case RESOLVED:
+            erasePositive(valPtr->payload.value);
+            break;
     }
 }
 
-/**
- * Here we require success and failure to be passed by reference (can be
- * stack-allocated). This is to work around an issue with the C ABI where
- * late arguments are scrambled.
- */
-void c_file_read(int32_t fd, struct Pos buffer, int64_t offset, struct Neg* success, struct Neg* failure) {
-
-    // Get the default loop
-    uv_loop_t* loop = uv_default_loop();
-
-    uint8_t* buffer_data = c_buffer_bytes(buffer);
-    int32_t len = (int32_t)c_buffer_length(buffer);
-
-    uv_buf_t buf = uv_buf_init((char*)buffer_data, len);
-
-    uv_fs_t* req = (uv_fs_t*)malloc(sizeof(uv_fs_t));
-
-    // Allocate Callbacks on the heap
-    Callbacks* callbacks = (Callbacks*)malloc(sizeof(Callbacks));
-    callbacks->on_success = *success;
-    callbacks->on_failure = *failure;
-
-    // // Store the callbacks in the req's data field
-    req->data = callbacks;
-
-    // // Argument `1` here means: we pass exactly one buffer
-    uv_fs_read(loop, req, fd, &buf, 1, offset, c_file_read_handler);
+void c_promise_resume_listeners(Listeners* listeners, struct Pos value) {
+	if (listeners != NULL) {
+		Stack head = listeners->head;
+		Listeners* tail = listeners->tail;
+		free(listeners);
+		c_promise_resume_listeners(tail, value);
+		sharePositive(value);
+		resume_Pos(head, value);
+	}
```

Comment:
Since when do we use tabs?

Reply:
Please check to not use tabs anywhere, this shows up in the whole file.

Reply:
Sorry!